examples/degrees.rs

```diff
@@ -0,0 +1,133 @@
+//! Simple DBSP example for counting the distribution of degrees in a graph.
+//!
+//! This program creates a series of graph edges, then counts the number of
+//! times each source node appears, then counts the number of times each count
+//! appears.  This is similar to the example for Differential Dataflow at
+//! <https://docs.rs/differential-dataflow/latest/differential_dataflow/>.
+
+use anyhow::Result;
+use clap::Parser;
+use dbsp::{
+    operator::FilterMap,
+    time::NestedTimestamp32,
+    trace::{BatchReader, Cursor},
+    DBData, DBWeight, OrdIndexedZSet, OutputHandle, Runtime,
+};
+
+type Node = usize;
+type Weight = isize;
+
+#[derive(Debug, Clone, Parser)]
+struct Args {
+    /// Number of initial edges in the graph.
+    #[clap(long, default_value = "100")]
+    edges: usize,
+
+    /// Number of source nodes in the graph.
+    #[clap(long, default_value = "13")]
+    sources: usize,
+
+    /// Number of extra edges added later to the graph.
+    #[clap(long, default_value = "5")]
+    extra: usize,
+
+    /// Number of threads.
+    #[clap(long, default_value = "2")]
+    threads: usize,
+}
+
+struct BatchIterator<'a, T: BatchReader> {
+    cursor: T::Cursor<'a>,
+}
+
+impl<'a, T: BatchReader> BatchIterator<'a, T> {
+    /// Returns an iterator of `(key, value, weight)` over the items that
+    /// `cursor` visits.
+    fn new(cursor: T::Cursor<'a>) -> Self {
+        Self { cursor }
+    }
+}
+
+impl<'a, K, V, R> Iterator for BatchIterator<'a, OrdIndexedZSet<K, V, R>>
```

Comment:
shouldn't this be in dbsp?

Reply:
I don't know. I don't understand why cursors aren't iterators.

Reply:
This is partially for historic reasons.  Before GATs were standardized, the cursor API was more complicated and couldn't be framed as an iterator.

Switching to an iterator-based API is worth exploring if we can do it without a performance penalty. There are several different ways to iterate over a batch. E.g., we may want to iterate over (key,value,weight) tuples as you do here or over keys with a nested iterator over (value, weight) tuples.

Reply:
No reason for a BatchAllKeyValuesIterator not to exist.